File: src/main/java/com/github/solairerove/algs4/leprosorium/arrays/LongestPeak.java
package com.github.solairerove.algs4.leprosorium.arrays;

/**
 * Find length of longest peak(adjacent increasing elements) in array.
 */
public class LongestPeak {
    public static void main(String[] args) {
        int[] a = {0, 1, 2, 3, 4, 4, -1, 12, 10, 6, 5, 4, 3, 2, -1, -3, 2, 4}; // 10
        System.out.println(longestPeak(a));
    }

    private static int longestPeak(int[] a) {
        int n = a.length;
        int longestPeak = 0, i = 1;

        while (i < n - 1) {
            var isPeak = a[i] > a[i - 1] && a[i] > a[i + 1];
            if (!isPeak) {
                i++;
                continue;
            }

            int leftIdx = i - 2, rightIdx = i + 2;
            while (leftIdx >= 0 && a[leftIdx] < a[leftIdx + 1]) leftIdx--;
            while (rightIdx < n && a[rightIdx] < a[rightIdx - 1]) rightIdx++;

            longestPeak = Math.max(longestPeak, rightIdx - leftIdx - 1);
            i = rightIdx;
        }

        return longestPeak;
    }
}

File: src/main/kotlin/com/github/solairerove/algs4/leprosorium/arrays/LongestPeak.kt
package com.github.solairerove.algs4.leprosorium.arrays

import kotlin.math.max

fun main() {
    val arr = listOf(3, -4, 7, 12, 4, 20, 15, 6, -2, 5, 16, 11, 1)
    print(longestPeak(arr)) // 5
}

// O(n) time | O(1) space
private fun longestPeak(arr: List<Int>): Int {
    var longestPeakLen = 0
    var i = 1

    while (i < arr.size - 1) {
        val isPeak = arr[i] > arr[i - 1] && arr[i] > arr[i + 1]

        if (!isPeak) {
            i++
            continue
        }

        var low = i - 2
        while (low >= 0 && arr[low] < arr[low + 1]) low--

        var high = i + 2
        while (high < arr.size && arr[high] < arr[high - 1]) high++

        val currPeakLen = high - low - 1
        longestPeakLen = max(longestPeakLen, currPeakLen)
        i = high
    }

    return longestPeakLen
}
